nixos/modules/services/admin/nomachine.nix

```diff
@@ -0,0 +1,233 @@
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to open port 4000 in the firewall.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.users = {
+      "nx" = {
+        group = "nx";
+        isSystemUser = true;
+        shell = "${nxserver}/NX/etc/nxserver";
+        home = "/var/lib/nxserver/nx";
+      };
+    };
+
+    users.groups = {
+      "nx" = {};
+    };
+
+    security.pam.services.nx.text = ''
+      # This is a default PAM configuration for NoMachine. It is based on
+      # system's 'su' configuration and can be adjusted freely according
+      # to administrative needs on the system.
+
+      auth       include       su
+      account    include       su
+      password   include       su
+      session    include       su
+    '';
+
+    security.pam.services.nxlimits.text = ''
+      # This is a default PAM configuration for NoMachine.
+      # Used to obtain nx user and nxhtd user limits.
+
+      session   optional pam_limits.so
+    '';
+
+    environment.etc = {
+      "NX/server/localhost/server.cfg" = {
+        text = ''ServerRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/server/localhost/node.cfg" = {
+        text = ''NodeRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/server/localhost/client.cfg" = {
+        text = ''ClientRoot = "${nxserver}/NX"'';
+        mode = "644";
+      };
+
+      "NX/nxserver".source = "${nxserver}/bin/nxserver";
+      "NX/nxnode".source = "${nxserver}/bin/nxnode";
+
+      "NX/server.cfg".text = serverConf;
+      "NX/node.cfg".text = nodeConf;
+    };
+
+    security.wrappers = {
+      nxexec = {
+        source = "${nxserver}/NX/bin/nxexec.orig";
+        owner = "root";
+        group = "root";
+        setuid = true;
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) [4000];
+
+    systemd.services.nxserver = {
+      description = "NoMachine Server daemon";
+      wantedBy = ["multi-user.target"];
+      after = ["syslog.target" "network.target" "network-online.target" "display-manager.service"];
+      wants = ["network-online.target"];
+      bindsTo = ["display-manager.service"];
+
+      serviceConfig = {
+        User = "nx";
+        Group = "nx";
+
+        LogsDirectory = "nxserver";
+
+        StateDirectory = [
+          "nxserver"
+          "nxserver/db"
+          "nxserver/db/server"
+          "nxserver/nx"
+        ];
+
+        RuntimeDirectory = [
+          "nxserver"
+          "nxserver/run"
+          "nxserver/tmp"
+          "nxserver/scripts"
+        ];
+
+        ExecStartPre = let
+          preStartScript = ''
+            set -euo pipefail
+
+            # always update to current version of files
+            cp ${nxserver}/NX/etc.static/version /etc/NX/
+            cp ${nxserver}/NX/etc.static/update.cfg /etc/NX/
+            cp ${nxserver}/NX/etc.static/usb.db /etc/NX/
```

Comment:
Is this mutable state? Isnt it better to copy only if the destination does not exist?

Reply:
Yeah, I think it's only the files from `nxclient` (`version` and `update.cfg`) that should be overwritten every time.

Reply:
I added a check for `usb.db`, the other files should to be updated after a nomachine update, so instead of adding logic to do that I think simply overwriting them is easier.